### PR TITLE
feat(api): refactor authentication drivers + manage github merge commit

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18474.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18474.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

File: engine/api/user_link.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+		}
+		defer tx.Rollback() // nolint
+
+		userLink := sdk.UserLink{
+			AuthentifiedUserID: u.ID,
+			Created:            time.Now(),
+			Username:           userInfo.Username,
+			Type:               string(consumerType),
+		}
+		log.Warn(ctx, "%+v", userLink)
```
Comment (yesnault) [inline engine/api/user_link.go:96]: debug / info?

**Review by yesnault — CHANGES_REQUESTED** [review]

File: engine/api/link/gorp.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	"github.com/ovh/cds/sdk"
+)
+
+type dbUserLink struct {
+	sdk.UserLink
+	gorpmapper.SignedEntity
+}
+
+func (ul dbUserLink) Canonical() gorpmapper.CanonicalForms {
+	_ = []interface{}{ul.ID, ul.AuthentifiedUserID, ul.Type, ul.Username}
```
Comment (richardlt) [inline engine/api/link/gorp.go:15]: This shoudl include external_id

File: sdk/token.go [hunk 3]
```diff
@@ -646,4 +655,5 @@ type AuthSigninConsumerToken struct {
 	RedirectURI       string `json:"redirect_uri,omitempty"`
 	RequireMFA        bool   `json:"require_mfa,omitempty"`
 	IsFirstConnection bool   `json:"is_first_connection,omitempty"`
+	LinkUser          bool   `json:"link_user,omitempty"`
```
Comment (richardlt) [inline sdk/token.go:658]: Why no creating dedicated payload for link requests ?

File: engine/api/api_routes.go [hunk 4]
```diff
@@ -380,6 +383,8 @@ func (api *API) InitRouter() {
 	r.Handle("/user/{permUsernamePublic}", Scope(sdk.AuthConsumerScopeUser), r.GET(api.getUserHandler), r.PUT(api.putUserHandler), r.DELETE(api.deleteUserHandler))
 	r.Handle("/user/{permUsernamePublic}/group", Scope(sdk.AuthConsumerScopeUser), r.GET(api.getUserGroupsHandler))
 	r.Handle("/user/{permUsername}/contact", Scope(sdk.AuthConsumerScopeUser), r.GET(api.getUserContactsHandler))
+	r.Handle("/user/{permUsername}/{consumerType}/link/callback", Scope(sdk.AuthConsumerScopeUser), r.POST(api.postLinkExternalUserWithCDSHandler))
```
Comment (richardlt) [inline engine/api/api_routes.go:386]: Why not a post on /link/{consumerType} ? I don't think that this route can be called be someone else than the current authentified user for its own account. Like the /auth/consumer/{consumerType}/signin route.

**Review by richardlt — CHANGES_REQUESTED** [review]

File: sdk/token.go (hunk 3 above)
Comment (sguiheux) [reply sdk/token.go:658]: In UI, it's the same page that handle the request. I though it's easier to manage and avoid field conflicts

**Review by yesnault — APPROVED** [review]

**Review by richardlt — APPROVED** [review]

**Review by fsamin — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18519.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18520.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Pull request merged**